exercises/practice/react/.meta/src/reference/java/React.java

```diff
@@ -0,0 +1,124 @@
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class React {
+
+    interface Cell<T> {
+        T getValue();
+        default T getTentativeValue() {
+            return getValue();
+        }
+        void addDependent(ComputeCell<T> dependent);
```

Comment:
To be honest I don't think we should encourage students to add methods to an interface that are meant for internal use only. I would leave the interfaces as-is compared to the starter implementation.

Reply:
I think this exercise can't be solved without adding some kind of internal method because of `testCallbacksShouldOnlyBeCalledOnceEvenIfMultipleDependenciesChange`. The test requires the `ComputeCell` to tell when a value change has finished propagating to know when it can finally fire the callbacks.

I think it is possible to solve this by adding the internal method to just `ComputeCellImpl`, but then that would require the implementing classes to hold a `Collection<ComputeCellImpl>` instead of `Collection<ComputeCell>` which doesn't feel right to me. I wonder if we should use just classes instead of interfaces which would allow the internal methods to be marked private or package private.

I took a look at other this is solved in other tracks (namely from the [Kotlin](https://github.com/exercism/kotlin/blob/main/exercises/practice/react/.meta/src/reference/kotlin/React.kt) and [Crystal](https://github.com/exercism/crystal/blob/main/exercises/practice/react/.meta/src/example.cr) tracks). I think their approach is a lot neater and will help avoid adding `getTentativeValue` on the base class/interface. I plan to try changing to this approach in the next commit.

Reply:
> I think it is possible to solve this by adding the internal method to just ComputeCellImpl, but then that would require the implementing classes to hold a `Collection<ComputeCellImpl>` instead of `Collection<ComputeCell>` which doesn't feel right to me.

That was exactly the direction I was going for. I'm considering the interfaces the 'public' API for the exercise, and the classes implementing them could even be private. I see no reason why these classes should not contain references to each other in private fields.